src/Middleware/HttpCache.php

```diff
@@ -0,0 +1,197 @@
+<?php
+namespace Yiisoft\Yii\Web\Middleware;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Router\Method;
+
+/**
+ * HttpCache implements client-side caching by utilizing the `Last-Modified` and `ETag` HTTP headers.
+ */
+final class HttpCache implements MiddlewareInterface
+{
+    private const DEFAULT_HEADER = 'public, max-age=3600';
+
+    /**
+     * @var callable a PHP callback that returns the UNIX timestamp of the last modification time.
+     * The callback's signature should be:
+     *
+     * ```php
+     * function ($request, $params)
+     * ```
+     *
+     * where `$request` is the [[ServerRequestInterface]] object that this filter is currently handling;
+     * `$params` takes the value of [[params]]. The callback should return a UNIX timestamp.
+     *
+     * @see http://tools.ietf.org/html/rfc7232#section-2.2
+     */
+    private $lastModified;
+
+    /**
+     * @var callable a PHP callback that generates the ETag seed string.
+     * The callback's signature should be:
+     *
+     * ```php
+     * function ($request, $params)
+     * ```
+     *
+     * where `$request` is the [[ServerRequestInterface]] object that this filter is currently handling;
+     * `$params` takes the value of [[params]]. The callback should return a string serving
+     * as the seed for generating an ETag.
+     */
+    private $etagSeed;
+
+    /**
+     * @var bool whether to generate weak ETags.
+     *
+     * Weak ETags should be used if the content should be considered semantically equivalent, but not byte-equal.
+     *
+     * @see http://tools.ietf.org/html/rfc7232#section-2.3
+     */
+    private bool $weakEtag = false;
+
+    /**
+     * @var mixed additional parameters that should be passed to the [[lastModified]] and [[etagSeed]] callbacks.
+     */
+    private $params;
+
+    /**
+     * @var string the value of the `Cache-Control` HTTP header. If null, the header will not be sent.
+     * @see http://tools.ietf.org/html/rfc2616#section-14.9
+     */
+    private ?string $cacheControlHeader = self::DEFAULT_HEADER;
+
+    private ResponseFactoryInterface $responseFactory;
+
+    public function __construct(ResponseFactoryInterface $responseFactory)
+    {
+        $this->responseFactory = $responseFactory;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $method = $request->getMethod();
+        if (
+            !\in_array($method, [Method::GET, Method::HEAD])
+            || ($this->lastModified === null && $this->etagSeed === null)
```

Comment:
In my opinion, these are mandatory data (`lastModified` and `etagSeed`). If not configured, it can be left out of the middleware stack.

Reply:
I don't understand. What did you mean?

Reply:
One of these is mandatory, not both. So the current code is OK.